### PR TITLE
Special max

### PR DESCRIPTION
**Pull request by beauxq**
a special range option to display a different label if the max is chosen

a feature I plan to use in the game I'm working on

### CODE REVIEW

**Comment by KonoTyran** [comment]
I feel like this could be generalized a bit more. Allow any specific number in your range to have a special label instead of just "max" this could be useful in other games as well then, example: Hollow Knights `Randomize Charm Notch Costs` -1 and -2 are special options that do different things and this could replace it with specific text

**Comment by beauxq** [comment]
I thought of other values, but didn't think of any other uses. Since people have thought of other uses, I'll work on that.

File: WebHostLib/static/assets/player-settings.js
Comment (LegendaryLinux) [inline WebHostLib/static/assets/player-settings.js:149]: You must check for the existence of this special property, as it is not guaranteed to be present on normal range options.

File: WebHostLib/static/assets/player-settings.js
Comment (LegendaryLinux) [inline WebHostLib/static/assets/player-settings.js:154]: This will cause the layout of the page to change suddenly and without warning. This behavior is undesirable.

File: WebHostLib/static/assets/player-settings.js
Comment (LegendaryLinux) [inline WebHostLib/static/assets/player-settings.js:151]: There is no need to assign this to memory, as it is already in memory.

File: WebHostLib/static/assets/player-settings.js
Comment (LegendaryLinux) [inline WebHostLib/static/assets/player-settings.js:153]: This evaluation uses a quirk of JS to perform assignment, and should be avoided.
Also, comparison should always use `===`.

**Review by LegendaryLinux — CHANGES_REQUESTED** [review]
The Python stuff looks fine to me, but the JS stuff needs some work. The page layout is especially concerning, since the length of the special text can cause sudden, rapid changes to the page styles.

File: WebHostLib/static/assets/player-settings.js
Comment (beauxq) [reply WebHostLib/static/assets/player-settings.js:153]: `===` doesn't work because the value from html is a string and settingMax is a number.

File: WebHostLib/static/assets/player-settings.js
Comment (beauxq) [reply WebHostLib/static/assets/player-settings.js:149]: That check is done where it is referenced. Should I also check it here if it's checked later?

Comment (LegendaryLinux) [reply WebHostLib/static/assets/player-settings.js:149]: If the property doesn't exist, attempting to access it will result in an an Error being thrown. You should ensure it exists before attempting to access it for the first time, which happens during assignment here. When using it later, you won't need to check for its existence again.

File: WebHostLib/static/assets/player-settings.js
Comment (LegendaryLinux) [reply WebHostLib/static/assets/player-settings.js:154]: First, I consider this discovery to be a bug report and I will be addressing it.

Second, sudden layout changes are undesirable and should be avoided.

Third, and in my opinion most importantly, those range sliders are intended for use with numbers only. Words were never supposed to be used in place of those numbers.

File: WebHostLib/static/assets/player-settings.js
Comment (LegendaryLinux) [reply WebHostLib/static/assets/player-settings.js:151]: To my knowledge, accessing the property of an object already in memory is not slower than assigning a new scalar. Using `event.target.value` should be both faster and use less memory. If you have documentation which shows otherwise, I will consider it.

If you perform some calculation or operation on data which should be referenced multiple times, that is worth assigning new memory for.

As for shorter expressions, that depends on the person reading the code. As the senior developer at my company, with ten years of professional experience, it has become my opinion that shortcut variables and short lines of code with the sole purpose of being short are undesirable and cause longer code analysis times (takes longer to read). I can provide further explanation and examples of this if you'd like.

File: WebHostLib/static/assets/player-settings.js
Comment (beauxq) [reply WebHostLib/static/assets/player-settings.js:154]: Do you have any better ideas for how to have an option that can be from a large range of numbers and one or a few other non-number values?

Multiple people have said they would like this, and Berserker suggested that I implement this.

File: WebHostLib/static/assets/player-settings.js
Comment (LegendaryLinux) [reply WebHostLib/static/assets/player-settings.js:153]: In that case, I'd recommend using `parseInt(event.target.value, 10)`.

Referencing an above comment, if you wanted to do something like: `const val = parseInt(event.target.value, 10);` and reference `val` more than once later, that would be proper.

As an aside, single letter variable names should also generally be avoided unless they are looping variables.

**Comment by LegendaryLinux** [comment]
This feature request is going to require me to consult with the designer before being approved. I'll need to come up with a new type of range option for both the player-settings and weighted-settings pages. This is likely to take a while.

**Note for myself later, this feature request is justified because it allows a developer to indicate extra significance for a specific value in a range.

File: WebHostLib/static/assets/player-settings.js
Comment (beauxq) [reply WebHostLib/static/assets/player-settings.js:151]: I am interested in further explanation and examples.

> with the sole purpose of being short

The purpose is not to be short. The purpose is to be easier to understand.
I agree that the single letter variable name was a bad choice, but it should still be in separate variable, just with a better name.

It only takes longer to read if you don't care about understanding the code.
It simplifies things to separate the ideas of the value and where the value came from.
`event.target.value` contains a lot of information about where the value came from that is not relevant where the value is used.
That information would be a distraction that makes the expression more difficult to understand.
`event` isn't relevant when we use the value. `target` isn't relevant when we use the value.

It also makes it easier to debug and maintain if there are more lines for breakpoints, and more places to insert debug messages, or insert the new feature that we're implementing.

Using `event.target.value` is not faster and I doubt it uses less memory.

The interpreters are optimized for this because many people recognize the value in writing code like this.

I've done my own tests to make sure this
```
        const x = obj.x;
        const y = x[0];
        const z = y.z;
        return z;
```
is just as fast as this
```
        return obj.x[0].z;
```

File: WebHostLib/static/assets/player-settings.js
Comment (LegendaryLinux) [reply WebHostLib/static/assets/player-settings.js:154]: As indicated in my final comment on the PR, I will be working on a new `range` control to accommodate for this.

File: WebHostLib/static/assets/player-settings.js
Comment (LegendaryLinux) [reply WebHostLib/static/assets/player-settings.js:151]: > `event.target.value` contains a lot of information about where the value came from that is not relevant where the value is used.

A value and its source are never more relevant than where they are being used. I don't understand this logic.

> That information would be a distraction that makes the expression more difficult to understand.

This argument would hold more weight in a discussion about literature, I think. If I was reading through a section of code, saw an assignment, and wanted to know where that assignment was made, I would need to look up the value's source. Using `event.target.value` in this case would prevent me from needing to look that information up.

In any case, as I have addressed, there is actually a computationally beneficial reason to use an assignment here. You're wanting to make a comparison, and should be using `===` to do so. To accomplish this, `parseInt` is required to ensure type consistency. Therefore, assigning the result of `parseInt(event.target.value, 10)` to your new `const` is preferable, and prevents repeat computation.

Comment (beauxq) [reply WebHostLib/static/assets/player-settings.js:151]: > A value and its source are never more relevant than where they are being used.

`event` isn't relevant when we use the value. `target` isn't relevant when we use the value.

You're claiming that they're relevant. How are they relevant?

You're presenting a hypothetical situation where you "want to know where that assignment was made".
That's begging the question. For the understanding of that expression, "where that assignment was made" is not relevant.

**Comment by LegendaryLinux** [comment]
Just to update on this, I haven't heard from the designer in a while, but I just reached out to her. I have an idea about how the new UI module might look, though.

**Comment by beauxq** [comment]
I was meaning to introduce the python data structure together with the web interface, so I was waiting on the interface decision.

But it looks like Berserker decided to work on the python data structure separately. https://github.com/ArchipelagoMW/Archipelago/pull/630

So when the web interface decision is made, it can be applied to that.

**Pull request closed without merging**
